File: frontend/src/lib/api/ml.ts
import { fetchClient } from './fetchClient';

export interface MlTargetCandidate {
  name: string;
  display_name: string;
  recommended_task: "classification" | "regression" | null;
  alternative_task: "classification" | "regression" | null;
  reason: string | null;
  non_null_count: number;
  missing_count: number;
  unique_count: number;
  is_eligible: boolean;
  exclusion_reason: string | null;
}

export interface MlFeatureRole {
  name: string;
  display_name: string;
  role: string;
  type: string;
  feature_status: "recommended" | "optional" | "excluded";
  selected_by_default: boolean;
  reason: string;
  warning: string | null;
}

export interface MLMetadataResponse {
  dataset_name: string;
  view: "mapped" | "working";
  has_cleaning_plan: boolean;
  row_count: number;
  target_candidates: MlTargetCandidate[];
  data_sufficiency_warnings: string[];
}

export interface MLValidateRequest {
  view: "mapped" | "working";
  target_column: string;
  task_type: "classification" | "regression";
  selected_features: string[] | null;
  test_size: number;
}

export interface MLValidationIssue {
  code: string;
  severity: string;
  message: string;
  class_label?: string;
  actual?: number;
  required?: number;
}

export interface MLValidateResponse {
  can_train: boolean;
  validation_warnings: string[];
  validation_issues: MLValidationIssue[];
  leakage_warnings: Array<{
    feature: string;
    severity: string;
    evidence: string;
    action_taken: string;
    explanation?: string;
  }>;
  target_statistics: {
    row_count: number;
    missing_count: number;
    unique_count: number;
    distribution?: Record<string, number>;
    class_distribution?: Record<string, number>;
    num_classes?: number;
    smallest_class_label?: string;
    smallest_class_count?: number;
    min?: number;
    max?: number;
    mean?: number;
  };
  estimated_training_size: number;
  estimated_test_size: number;
  recommended_features_meta?: MlFeatureRole[];
  optional_features_meta?: MlFeatureRole[];
  excluded_features_meta?: MlFeatureRole[];
  default_selected_features?: string[];
}

export interface MLTrainRequest {
  view: "mapped" | "working";
  target_column: string;
  task_type: "classification" | "regression";
  selected_features: string[];
  test_size: number;
  models: string[];
}

export interface MLModelCandidateResult {
  model_id: string;
  display_name: string;
  is_baseline: boolean;
  status: string;
  cv_mean?: number;
  cv_std?: number;
  cv_min?: number;
  cv_max?: number;
  cv_fold_scores?: number[];
  test_metric?: number;
  primary_metric_name: string;
  metric_direction?: string;
  training_duration_seconds: number;
  failure_reason?: string;
}

export interface MLExperimentSummary {
  id: string;
  dataset_id: string;
  dataset_name: string;
  dataset_view: string;
  target_column: string;
  task_type: string;
  status: string;
  best_model_name?: string;
  primary_metric?: string;
  best_cv_metric?: number;
  test_metric?: number;
  created_at: string;
  completed_at?: string;
  artifact_available: boolean;
}

export interface MLExperimentResponse {
  id: string;
  workspace_id: string;
  dataset_id: string;
  dataset_view: string;
  target_column: string;
  task_type: string;
  status: string;
  selected_features: string[];
  excluded_features: string[];
  primary_metric?: string;
  best_model_name?: string;
  baseline_metric?: number;
  best_cv_metric?: number;
  test_metric?: number;
  candidate_results?: MLModelCandidateResult[];
  classification_evaluation?: Record<string, unknown>;
  regression_evaluation?: Record<string, unknown>;
  feature_importance?: Array<{feature: string, importance: number}>;
  feature_importance_method?: string;
  prediction_schema?: Array<{
    name: string;
    display_name: string;
    role: string;
    input_type: 'integer' | 'decimal' | 'select' | 'boolean' | 'date' | 'text';
    required?: boolean;
    allows_missing?: boolean;
    categories?: string[] | null;
    minimum?: number | null;
    maximum?: number | null;
    step?: number | null;
  }> | null;
  row_count?: number;
  training_row_count?: number;
  test_row_count?: number;
  created_at: string;
  completed_at?: string;
  error_message?: string;
  legacy_preprocessing_warnings?: Array<{
    code: string;
    feature: string;
    message: string;
  }>;
  artifact_available: boolean;
}

export interface MLPredictRequest {
  features: Record<string, unknown>;
}

export interface MLClassProbability {
  label: string;
  probability: number;
}

export interface MLPredictResponse {
  task_type: string;
  prediction: unknown;
  probabilities?: MLClassProbability[] | null;
  maximum_probability?: number | null;
  low_confidence?: boolean;
  confidence_message?: string | null;
  limitation?: string;
  input_validation_warnings: string[];
}

function mlDatasetBase(workspaceId: string, datasetId: string): string {
  return `workspaces/${encodeURIComponent(workspaceId)}/datasets/${encodeURIComponent(datasetId)}/ml`;
}

function mlExperimentBase(workspaceId: string, experimentId: string): string {
  return `workspaces/${encodeURIComponent(workspaceId)}/ml/experiments/${encodeURIComponent(experimentId)}`;
}

export const MLService = {
  getMetadata: async (workspaceId: string, datasetId: string, view: string = "original"): Promise<MLMetadataResponse> => {
    return fetchClient(`${mlDatasetBase(workspaceId, datasetId)}/metadata?view=${view}`);
  },

  validateConfig: async (workspaceId: string, datasetId: string, req: MLValidateRequest): Promise<MLValidateResponse> => {
    return fetchClient(`${mlDatasetBase(workspaceId, datasetId)}/validate`, {
      method: 'POST',
      body: JSON.stringify(req),
    });
  },

  train: async (workspaceId: string, datasetId: string, req: MLTrainRequest): Promise<MLExperimentResponse> => {
    return fetchClient(`${mlDatasetBase(workspaceId, datasetId)}/train`, {
      method: 'POST',
      body: JSON.stringify(req),
    });
  },

  listExperiments: async (workspaceId: string, datasetId?: string): Promise<MLExperimentSummary[]> => {
    let url = `workspaces/${encodeURIComponent(workspaceId)}/ml/experiments`;
    if (datasetId) {
      url += `?dataset_id=${encodeURIComponent(datasetId)}`;
    }
    return fetchClient(url);
  },

  getExperiment: async (workspaceId: string, experimentId: string): Promise<MLExperimentResponse> => {
    return fetchClient(mlExperimentBase(workspaceId, experimentId));
  },

  deleteExperiment: async (workspaceId: string, experimentId: string): Promise<void> => {
    return fetchClient(mlExperimentBase(workspaceId, experimentId), {
      method: 'DELETE',
    });
  },

  predict: async (workspaceId: string, experimentId: string, features: Record<string, unknown>): Promise<MLPredictResponse> => {
    return fetchClient(`${mlExperimentBase(workspaceId, experimentId)}/predict`, {
      method: 'POST',
      body: JSON.stringify({ features }),
    });
  }
};

File: backend/app/schemas/ml.py
from typing import List, Dict, Any, Optional, Literal
from pydantic import BaseModel, Field

class MLTargetStats(BaseModel):
    row_count: int
    missing_count: int
    unique_count: int
    distribution: Optional[Dict[str, int]] = None
    min: Optional[float] = None
    max: Optional[float] = None
    mean: Optional[float] = None
    class_distribution: Optional[Dict[str, int]] = None
    num_classes: Optional[int] = None
    smallest_class_label: Optional[str] = None
    smallest_class_count: Optional[int] = None

class MLValidationIssue(BaseModel):
    code: str
    severity: str
    message: str
    class_label: Optional[str] = None
    actual: Optional[float] = None
    required: Optional[float] = None


class MLTargetCandidate(BaseModel):
    name: str
    display_name: str
    recommended_task: Optional[str] = None
    alternative_task: Optional[str] = None
    reason: Optional[str] = None
    non_null_count: int
    missing_count: int
    unique_count: int
    is_eligible: bool
    exclusion_reason: Optional[str] = None

class MLFeatureRole(BaseModel):
    name: str
    display_name: str
    role: str
    type: str
    feature_status: Literal["recommended", "optional", "excluded"]
    selected_by_default: bool
    reason: str
    warning: Optional[str] = None

class MLTaskRecommendation(BaseModel):
    recommended_task: str
    reason: str
    alternative_task: Optional[str] = None
    can_override: bool = False
    warnings: List[str] = []

class MLMetadataResponse(BaseModel):
    dataset_name: str
    view: str
    has_cleaning_plan: bool
    row_count: int
    target_candidates: List[MLTargetCandidate]
    data_sufficiency_warnings: List[str]

class MLValidateRequest(BaseModel):
    view: Literal["mapped", "working"]
    target_column: str
    task_type: str
    selected_features: Optional[List[str]] = None
    test_size: float = Field(default=0.2)

class MLValidateResponse(BaseModel):
    task_type: str
    included_features: List[str]
    excluded_features: List[str]
    leakage_warnings: List[Dict[str, Any]]
    target_statistics: MLTargetStats
    estimated_training_size: int
    estimated_test_size: int
    validation_warnings: List[str]  # Deprecated
    validation_issues: List[MLValidationIssue] = []
    can_train: bool
    
    # Optional recommendation fields when requested
    recommended_features_meta: Optional[List[MLFeatureRole]] = None
    optional_features_meta: Optional[List[MLFeatureRole]] = None
    excluded_features_meta: Optional[List[MLFeatureRole]] = None
    default_selected_features: Optional[List[str]] = None

class MLTrainRequest(BaseModel):
    view: Literal["mapped", "working"]
    target_column: str
    task_type: str
    selected_features: List[str]
    test_size: float = Field(default=0.2)
    
class MLModelCandidateResult(BaseModel):
    model_id: str
    display_name: str
    is_baseline: bool
    status: str
    cv_mean: Optional[float] = None
    cv_std: Optional[float] = None
    cv_min: Optional[float] = None
    cv_max: Optional[float] = None
    cv_fold_scores: Optional[List[float]] = None
    test_metric: Optional[float] = None
    primary_metric_name: str
    metric_direction: Optional[str] = None
    training_duration_seconds: float
    failure_reason: Optional[str] = None

class MLFeatureImportance(BaseModel):
    feature: str
    original_feature: Optional[str] = None
    importance: float
    rank: int
    direction: Optional[str] = None # positive or negative

class MLPredictionSchemaItem(BaseModel):
    name: str
    display_name: str
    role: str
    input_type: Literal["integer", "decimal", "select", "boolean", "date", "text"]
    required: bool = False
    allows_missing: bool = True
    categories: Optional[List[str]] = None
    minimum: Optional[float] = None
    maximum: Optional[float] = None
    step: Optional[float] = None

class MLClassDistribution(BaseModel):
    label: str
    count: int

class MLConfusionMatrix(BaseModel):
    labels: List[str]
    matrix: List[List[int]]
    normalized_matrix: List[List[float]]

class MLClassificationEvaluation(BaseModel):
    accuracy: float
    balanced_accuracy: float
    precision_weighted: float
    recall_weighted: float
    f1_weighted: float
    f1_macro: float
    per_class_metrics: Dict[str, Dict[str, float]]
    confusion_matrix: MLConfusionMatrix
    roc_curve_data: Optional[List[Dict[str, float]]] = None
    roc_auc: Optional[float] = None
    log_loss: Optional[float] = None
    class_distribution: List[MLClassDistribution]

class MLRegressionEvaluation(BaseModel):
    mae: float
    mse: float
    rmse: float
    r2: float
    explained_variance: float
    median_absolute_error: Optional[float] = None
    mape: Optional[float] = None
    actual_vs_predicted: Optional[List[Dict[str, Any]]] = None
    residuals: Optional[List[Dict[str, Any]]] = None
    residual_summary: Optional[Dict[str, Any]] = None
    target_range: Optional[Dict[str, float]] = None
    prediction_range: Optional[Dict[str, float]] = None

class MLExperimentSummary(BaseModel):
    id: str
    dataset_id: str
    dataset_name: str
    dataset_view: str
    target_column: str
    task_type: str
    status: str
    best_model_name: Optional[str] = None
    primary_metric: Optional[str] = None
    best_cv_metric: Optional[float] = None
    test_metric: Optional[float] = None
    created_at: str
    completed_at: Optional[str] = None
    artifact_available: bool
class LegacyPreprocessingWarning(BaseModel):
    code: str
    feature: str
    message: str


class MLExperimentResponse(BaseModel):
    id: str
    workspace_id: str
    dataset_id: str
    dataset_view: str
    target_column: str
    task_type: str
    status: str
    
    selected_features: List[str]
    excluded_features: List[str]
    
    primary_metric: Optional[str] = None
    best_model_name: Optional[str] = None
    baseline_metric: Optional[float] = None
    best_cv_metric: Optional[float] = None
    test_metric: Optional[float] = None
    
    candidate_results: Optional[List[MLModelCandidateResult]] = None
    
    classification_evaluation: Optional[MLClassificationEvaluation] = None
    regression_evaluation: Optional[MLRegressionEvaluation] = None
    
    feature_importance: Optional[List[MLFeatureImportance]] = None
    feature_importance_method: Optional[str] = None
    
    prediction_schema: Optional[List[MLPredictionSchemaItem]] = None
    legacy_preprocessing_warnings: List[LegacyPreprocessingWarning] = []
    
    row_count: Optional[int] = None
    training_row_count: Optional[int] = None
    test_row_count: Optional[int] = None
    
    error_message: Optional[str] = None
    warnings: List[str] = []
    
    created_at: str
    completed_at: Optional[str] = None
    artifact_available: bool = False

class MLPredictRequest(BaseModel):
    features: Dict[str, Any]

class MLClassProbability(BaseModel):
    label: str
    probability: float

class MLPredictResponse(BaseModel):
    task_type: str
    prediction: Any
    probabilities: Optional[List[MLClassProbability]] = None
    maximum_probability: Optional[float] = None
    low_confidence: Optional[bool] = False
    confidence_message: Optional[str] = None
    limitation: Optional[str] = "Predicted probabilities are model estimates and are not guarantees."
    input_validation_warnings: List[str] = []
